File: 530-minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void inorder_traversel(TreeNode* root, vector<int> &arr) {
        if(root == NULL) return;
        inorder_traversel(root->left, arr);
        arr.push_back(root->val);
        inorder_traversel(root->right, arr);
    }
public:
    int getMinimumDifference(TreeNode* root) {
        //binary search tree will give sorted values in inorder traversal
        vector<int> arr;
        inorder_traversel(root, arr);
        int n = arr.size();
        int minValue = INT_MAX;

        for(int i = 0; i < n-1; i++) {
            minValue = min(minValue, arr[i+1] - arr[i]);
        }

        return minValue;
    }
};

File: 530-minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inorder_traversal(TreeNode root, ArrayList<Integer> arr) {
        if(root == null) return;
        inorder_traversal(root.left, arr);
        arr.add(root.val);
        inorder_traversal(root.right, arr);
    }
    public int getMinimumDifference(TreeNode root) {
        ArrayList<Integer> arr = new ArrayList<>();
        inorder_traversal(root, arr);
        int n = arr.size();
        int minValue = Integer.MAX_VALUE;
        for(int i = 0; i < n-1; i++) {
            minValue = Math.min(minValue, arr.get(i+1) - arr.get(i));
        }
        return minValue;
    }
}

